core/blockchain/production/impl/block_producer_impl.cpp

```diff
@@ -0,0 +1,151 @@
+          });
+    }
+    OUTCOME_TRY(bls_aggregate_sign,
+                bls_provider_->aggregateSignatures(bls_signatures));
+    Time now = clock_->nowUTC();
+    OUTCOME_TRY(current_epoch, epoch_->epochAtTime(now));
+    BlockHeader header{
+        .miner = std::move(miner_address),
+        .ticket = ticket,
+        .epost_proof = std::move(proof),
+        .parents = std::move(parent_tipset.cids),
+        .parent_weight = parent_weight,
+        .height = static_cast<uint64_t>(current_epoch),
+        .parent_state_root = std::move(vm_result.state_root),
+        .parent_message_receipts = std::move(vm_result.message_receipts),
+        .messages = msg_meta.getCID(),
+        .bls_aggregate = {bls_aggregate_sign.begin(), bls_aggregate_sign.end()},
+        .timestamp = static_cast<uint64_t>(now.unixTime().count()),
+        .block_sig = {},  // Block must be signed be Actor Miner
+        .fork_signaling = 0};
+    Block block{.header = std::move(header),
+                .bls_messages = std::move(bls_messages),
+                .secp_messages = std::move(secp_messages)};
+    return std::move(block);
+  }
+
+  outcome::result<BlockProducerImpl::Tipset> BlockProducerImpl::getTipset(
+      const CID &tipset_id) const {
+    auto raw_data = data_storage_->get(tipset_id);
+    if (raw_data.has_error()) {
+      return BlockProducerError::PARENT_TIPSET_NOT_FOUND;
+    }
+    gsl::span<const uint8_t> raw_span{raw_data.value().toVector()};
+    auto tipset = codec::cbor::decode<Tipset>(raw_span);
+    if (tipset.has_error()) {
+      return BlockProducerError::PARENT_TIPSET_INVALID_CONTENT;
+    }
+    return tipset.value();
+  }
+
+  outcome::result<BlockProducerImpl::MsgMeta>
+  BlockProducerImpl::getMessagesMeta(
+      const std::vector<SignedMessage> &messages) {
+    auto bls_backend = std::make_shared<InMemoryDatastore>();
+    auto secp_backend = std::make_shared<InMemoryDatastore>();
+    Amt bls_messages_amt{bls_backend};
+    Amt secp_messages_amt{secp_backend};
+    for (size_t index = 0; index < messages.size(); ++index) {
+      const SignedMessage &msg = messages.at(index);
+      BOOST_OUTCOME_TRY_INVOKE_TRY1(visit_in_place(
+          msg.signature,
+          [index, &msg, &bls_messages_amt](const BlsSignature &signature) {
+            return bls_messages_amt.setCbor(index, msg);
+          },
+          [index, &msg, &secp_messages_amt](
+              const Secp256k1Signature &signature) {
+            return secp_messages_amt.setCbor(index, msg);
+          }));
+    }
+    OUTCOME_TRY(bls_root_cid, bls_messages_amt.flush());
+    OUTCOME_TRY(bls_root, bls_backend->getCbor<Root>(bls_root_cid));
+    OUTCOME_TRY(secp_root_cid, secp_messages_amt.flush());
+    OUTCOME_TRY(secp_root, secp_backend->getCbor<Root>(secp_root_cid));
+    MsgMeta msg_meta{};
+    msg_meta.bls_messages = bls_root.getCID();
+    msg_meta.secpk_messages = secp_root.getCID();
+    return msg_meta;
```

Comment:
`bls_root_cid` and `secp_root_cid` are already cids.
Use `return MsgMeta{.bls_messages=bls_root_cid, .secp_messages=secp_root_cid}`

Reply:
Fixed, but in current implementation direct initialization of the MsgMeta is unavailable